proseco/gui.py

```diff
@@ -0,0 +1,375 @@
+    spoilslope = STAR_CHAR['General']['Spoiler']['Slope']
+    nSigma = opt['Spoiler']['SigErrMultiplier']
+    magdifflim = STAR_CHAR['General']['Spoiler']['MagDiffLimit']
+    if magdifflim == '-_Inf_':
+        magdifflim = -1 * np.inf
+
+    # If this check has really already been done for this nSigma, exclude the stars
+    # as candidates from that previous check.  However, there could be other candidates
+    # in the list that are new (relaxed constraints in stages) so continue to redo the search on
+    # for mag spoiled stars on all still-valid candidates.
+    mag_col = 'mag_spoiled_{}'.format(nSigma)
+    mag_spoil_check = 'mag_spoil_check_{}'.format(nSigma)
+    if mag_col in stars.columns:
+        ok = ok & ~stars[mag_spoil_check]
+
+    # If there are already no candidates, there isn't anything to do
+    if not np.any(ok):
+        return np.zeros_like(ok), ok
+
+    # Search the sky around 'ok' stars to see if they are spoiled by other stars
+    coords = SkyCoord(stars['RA_PMCORR'], stars['DEC_PMCORR'],
+                      unit='deg')
+    maxsep_arcs = maxsep * PIX_2_ARC
+    cand_idx_in_ok, cat_idx, spoil_dist, dist3d = search_around_sky(
+        coords[ok],
+        coords,
+        seplimit=u.arcsec * maxsep_arcs)
+    # Index the candidates back into the full list so we have one
+    # index to work with, shared betweek cand_idx and cat_idx
+    cand_idx = np.flatnonzero(ok)[cand_idx_in_ok]
+    # Try to find the spoilers in a vectorized way
+    cand_mags = stars['MAG_ACA'][cand_idx]
+    spoiler_mags = stars['MAG_ACA'][cat_idx]
+    itself = cand_idx == cat_idx
+    too_dim = (cand_mags - spoiler_mags) < magdifflim
+    delmag = (cand_mags - spoiler_mags
+              + nSigma * np.sqrt(stderr2[cand_idx] + stderr2[cat_idx]))
+    thsep = intercept + delmag * spoilslope
+    spoils = (spoil_dist < u.arcsec * thsep * PIX_2_ARC) & ~itself & ~too_dim
+    # This is now indexed over the cat/cand idxs so, re-index again
+    spoiled = np.zeros_like(ok)
+    spoiled[np.unique(cand_idx[spoils])] = True
+    # Include any previous spoilers
+    spoiled = spoiled | stars[mag_col]
+    return spoiled, ok
+
+
+def check_bad_pixels(stars, not_bad, dither, opt):
+    """
+    Check if any candidate is spoiled by a bad pixel.
+    """
+    raw_bp = np.array(opt['Body']['Pixels']['BadPixels'])
+    bp_row = []
+    bp_col = []
+    # Bad pixel entries are [row_min, row_max, col_min, col_max].
+    # Convert this to lists of the row and col coords of the bad pixels
+    for row in raw_bp:
+        for rr in range(row[0], row[1] + 1):
+            for cc in range(row[2], row[3] + 1):
+                bp_row.append(rr)
+                bp_col.append(cc)
+    bp_row = np.array(bp_row)
+    bp_col = np.array(bp_col)
+
+    # Then for the pixel region of each candidate, see if there is a bad
+    # pixel in the region.
+    spoiled = np.zeros_like(not_bad)
+    row_extent = 4 + dither[0] * ARC_2_PIX
```

Comment:
Being conservative (but not hugely conservative), the one-shot error could be 80", which means a box with halfw = 80 + dither.  With that basically there will be difficulty choosing guide stars.

We can instead stipulate that the checking here relates to the stable guide phase where the attitude error is zeroed-out.  The assumption (a good one) is that the journey while correcting the maneuver error may hit some bumps on the road (hot/bad pixels) but not enough to prevent getting to the destination.

One thing to keep in mind is that for proseco 4.0 (initial release) the guide selection here does not need to be perfect.  The intent was to just fix a few low-hanging fruit from SAUSAGE and clean the code.  Does SAUSAGE worry about bad pixels in the maneuver error box?

Reply:
SAUSAGE does not worry about bad pixels in the maneuver error box.

And regarding "low-hanging fruit", the item on my list was "do something more reasonable for bad pixel checking".  *Considering* the possibility of initial error in pixel space seemed reasonable (which we've done here), and actually implementing a secondary preference to avoid bright pixels in a bigger region around the anticipated guide star region (via stages) would be trivial if we decided to go that way.

Appropriate testing, however, is difficult to define either way (with the simplest changes or with more complicated changes to the stages).

Reply:
I thought the "more reasonable" ended at mag-dependence for the existing check.

But certainly giving thought to this is a good thing.  My problem is always getting stuck with the whole stage formalism when you get to questions like what is preferred:
- 9.1 mag star with no nearby bad pix but a very bright pixel 58 arcsec away
- 9.7 mag star with no bad pix within 60 + 8 arcsec
- 9.0 mag star with an 10.5 mag bad pix imposter 25 arcsec away

Reply:
Well, in the other PR we've expanded to using the 2x2 pixel bin instead of the max pixel in the region (and I still have questions on that).

And regarding stanges, indeed, trivial to implement in code, non-trivial to decide upon!